finality-aleph/src/new_network/connection_manager/service.rs

```diff
@@ -0,0 +1,662 @@
+                (None, Vec::new())
+            }
+        }
+    }
+
+    /// Returns a data sink for data received that is associated with the given session.
+    pub fn get_session_channel(&self, session_id: &SessionId) -> Option<&mpsc::UnboundedSender<D>> {
+        self.sessions
+            .get(session_id)
+            .map(|session| session.data_for_user.as_ref())
+            .flatten()
+    }
+}
+
+/// Input/output interface for the connectiona manager service.
+pub struct IO<D: Clone + Codec> {
+    commands_for_network: mpsc::UnboundedSender<ConnectionCommand>,
+    messages_for_network: mpsc::UnboundedSender<(NetworkData<D>, DataCommand)>,
+    commands_from_user: mpsc::UnboundedReceiver<SessionCommand<D>>,
+    messages_from_user: mpsc::UnboundedReceiver<(D, SessionId, Recipient)>,
+    messages_from_network: mpsc::UnboundedReceiver<NetworkData<D>>,
+}
+
+#[derive(Debug)]
+enum SendError {
+    Network,
+    Command,
+    User,
+}
+
+impl<D: Clone + Codec> IO<D> {
+    fn send_data(&self, to_send: (NetworkData<D>, DataCommand)) -> Result<(), SendError> {
+        self.messages_for_network
+            .unbounded_send(to_send)
+            .map_err(|_| SendError::Network)
+    }
+
+    fn send_command(&self, to_send: ConnectionCommand) -> Result<(), SendError> {
+        self.commands_for_network
+            .unbounded_send(to_send)
+            .map_err(|_| SendError::Command)
+    }
+
+    fn send(
+        &self,
+        to_send: (
+            Option<ConnectionCommand>,
+            Vec<(NetworkData<D>, DataCommand)>,
+        ),
+    ) -> Result<(), SendError> {
+        if let Some(command) = to_send.0 {
+            self.send_command(command)?;
+        }
+        for data_to_send in to_send.1 {
+            self.send_data(data_to_send)?;
+        }
+        Ok(())
+    }
+
+    fn on_data<NI: NetworkIdentity>(
+        &self,
+        service: &Service<NI, D>,
+        data: D,
+        session_id: SessionId,
+    ) -> Result<(), SendError> {
+        if let Some(data_for_session_user) = service.get_session_channel(&session_id) {
+            data_for_session_user
+                .unbounded_send(data)
+                .map_err(|_| SendError::User)
+        } else {
+            debug!(target:"aleph-network", "Received data from unknown session {:?}", session_id);
+            Ok(())
+        }
+    }
+
+    fn on_network_message<NI: NetworkIdentity>(
+        &self,
+        service: &mut Service<NI, D>,
+        message: NetworkData<D>,
+    ) -> Result<(), SendError> {
+        use NetworkData::*;
+        match message {
+            Meta(message) => self.send(service.on_discovery_message(message)),
+            Data(data, session_id) => self.on_data(service, data, session_id),
+        }
+    }
+
+    /// Run the connection manager service with this IO.
+    pub async fn run<NI: NetworkIdentity>(mut self, mut service: Service<NI, D>) {
```

Comment:
Since this function looks like can fail in multiple number of ways consider changing it to `fn run...(...) -> Result<...>`

Reply:
Changed as suggested, this added a couple more errors, but in general might be more concise, so nice.